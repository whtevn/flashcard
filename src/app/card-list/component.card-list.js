import React from 'react'
import FCD_Flashcard from '../flashcard';

class FCD_CardList extends React.Component {
  render (){
    const cards = this.props.cards;
    const onSelect = this.props.onSelect;
    return (
      <div >
        { cards.map((card) => < FCD_Flashcard question={card.get('question')} answer={card.get('answer')} key={card.get('key')} /> ) }
      </div>
    )
  }
}
export default FCD_CardList;

