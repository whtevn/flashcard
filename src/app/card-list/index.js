import FCD_CardList from './container.card-list';
export default FCD_CardList;
export { Reducer, initialState } from './reducer.card-list';
