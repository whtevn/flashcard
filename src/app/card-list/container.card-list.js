import { fromJS } from 'immutable';
import { connect } from 'react-redux';
import { default as Component } from './component.card-list';
import * as Actions from './actions.card-list';

// connect component to store
export function mapStateToProps(state, ownprops){
  state = state.cards || state
  return {
    cards: state.getIn(['data', 'cards']),
    selected: state.getIn(['data', 'selected'])
  }
}

export function mapDispatchToProps(dispatch){
  return {
    onSelect: (selected, cards) => {
      dispatch(Actions.Select( fromJS({selected, cards}) ));
    }
  }
}

const FCD_CardList = connect(mapStateToProps, mapDispatchToProps)(Component);
export default FCD_CardList;

