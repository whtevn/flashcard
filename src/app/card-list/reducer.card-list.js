import { fromJS } from 'immutable';
import * as Actions from './actions.card-list'

export const initialState = fromJS({
  data: {
    cards: [],
    selected: undefined
  }
})

export const Reducer = (state = initialState, action) => {
  switch (action.type) {
    case Actions.SELECT:
      return state
              .setIn(['data', 'cards'], action.payload.get('cards'))
              .setIn(['data', 'results'], Map())
              .setIn(['data', 'selected'], 0);
    case Actions.CREATE:
      const cards = state.getIn(['data', 'cards']);
      return state.setIn(['data', 'cards'], cards.push(
        action.payload.get('card').merge(fromJS({
          key: cards.size
        }))))
    default:
      return state;
  }
}
