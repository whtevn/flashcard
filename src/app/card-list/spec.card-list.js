import renderer from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {default as Component} from './component.card-list';
import * as Container from './container.card-list';
import FCD_CardList, { initialState } from './';
import { List, Map, fromJS } from 'immutable';
import { Reducer } from './reducer.card-list';
import * as Actions from './actions.card-list';
const mockFn = jest.fn();

describe("The card list", ()=>{
  let cardList;
  let store;
  beforeEach(()=>{
    cardList = List()
                  .push(Map({question: "q 1", key: 0, answer: "a 1"}))
                  .push(Map({question: "q 2", key: 1, answer: "a 2"}))
                  .push(Map({question: "q 3", key: 2, answer: "a 3"}))
                  .push(Map({question: "q 4", key: 3, answer: "a 4"}))
                  .push(Map({question: "q 5", key: 4, answer: "a 5"}))
    store = createStore(Reducer, {cards: cardList});
  })

  describe("component", ()=>{
    test('Card list renders correctly', () => {
      const tree = renderer.create(
            <Provider store={store}>
              <MuiThemeProvider>
                <Component cards={cardList} onSelect={()=>{}} />
              </MuiThemeProvider>
            </Provider>
          ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  })

  describe("container", ()=>{
    let state, props, selected;
    beforeEach(()=>{
      selected = 0;
    })
    describe('mapping state to props', ()=>{
      beforeEach(()=>{
        state = { cards: fromJS({data: {cards: cardList, selected}}) };
        props = Container.mapStateToProps(state)
      })
      it("should set the cards", ()=>{
        expect(props.cards).toBe(cardList)
      })

      it("should set the selected", ()=>{
        expect(props.selected).toBe(selected)
      })
    })

    describe('mapping dispatch to props', ()=>{
      beforeEach(()=>{
        props = Container.mapDispatchToProps(mockFn)
        props.onSelect(selected, cardList)
      })
      afterEach(() => {
        mockFn.mockClear();
      })
      describe("when a card is selected", ()=>{
        it("should dispatch the select action", ()=>{
          expect(mockFn.mock.calls).toMatchSnapshot()
        })
      })
    })
  })

  xdescribe("reducer", ()=>{
    let state, action, next, selected;
    describe('CardList', ()=>{
      beforeEach(()=>{
        selected = 100;
        state = fromJS({data: {cards: cardList, selected: 0}});
        action = { type: Actions.SELECT, payload: fromJS({ selected }) }
        next  = Reducer(state, action);
      })
      describe("::SELECT", ()=>{
        it("should read the selected id from the payload and set it in the state", ()=>{
          expect(next.getIn(['data', 'selected'])).toBe(selected);
        })
      })
    })
  })
})

