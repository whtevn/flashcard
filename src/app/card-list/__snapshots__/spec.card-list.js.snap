// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The card list component Card list renders correctly 1`] = `
<div>
  <div>
    <div>
      <div>
        q 1
      </div>
    </div>
    <div>
      <div>
        a 1
      </div>
    </div>
  </div>
  <div>
    <div>
      <div>
        q 2
      </div>
    </div>
    <div>
      <div>
        a 2
      </div>
    </div>
  </div>
  <div>
    <div>
      <div>
        q 3
      </div>
    </div>
    <div>
      <div>
        a 3
      </div>
    </div>
  </div>
  <div>
    <div>
      <div>
        q 4
      </div>
    </div>
    <div>
      <div>
        a 4
      </div>
    </div>
  </div>
  <div>
    <div>
      <div>
        q 5
      </div>
    </div>
    <div>
      <div>
        a 5
      </div>
    </div>
  </div>
</div>
`;

exports[`The card list container mapping dispatch to props when a card is selected should dispatch the select action 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "cards": Array [
          Object {
            "answer": "a 1",
            "key": 0,
            "question": "q 1",
          },
          Object {
            "answer": "a 2",
            "key": 1,
            "question": "q 2",
          },
          Object {
            "answer": "a 3",
            "key": 2,
            "question": "q 3",
          },
          Object {
            "answer": "a 4",
            "key": 3,
            "question": "q 4",
          },
          Object {
            "answer": "a 5",
            "key": 4,
            "question": "q 5",
          },
        ],
        "selected": 0,
      },
      "type": "Card List SELECT",
    },
  ],
]
`;
